### DISCUSSION TITLE
Cannot find -lNatNet

### DISCUSSION THREAD

**Discussion by The-SS** [USC-ACTLab/crazyswarm#522] [Q&A] [answered]
Per [https://github.com/whoenig/NatNetSDKCrossplatform/issues/3](https://github.com/whoenig/NatNetSDKCrossplatform/issues/3), I cloned crazyswarm and pulled `crazyswarm/ros_ws/src/crazyswarm/externalDependencies/libmotioncapture` recursively. I changed [ENABLE_OPTITRACK_CLOSED_SOURCE](https://github.com/IMRCLab/libmotioncapture/blob/13d14f85d29171426a2b663cf71cb6262c6f7760/CMakeLists.txt#L8) to ON and was able to build `libmotioncapture`. 
However, when building with crazyswarm's `./build.sh` script, the package is failing to find NatNet.
```
[ 98%] Linking CXX executable /home/mobilerobot/Crazyflie_robots/crazyswarm/ros_ws/devel/lib/crazyswarm/crazyswarm_server
[ 98%] Linking CXX executable /home/mobilerobot/Crazyflie_robots/crazyswarm/ros_ws/devel/lib/crazyswarm/mocap_helper
/usr/bin/ld: cannot find -lNatNet
collect2: error: ld returned 1 exit status
make[2]: *** [crazyswarm/CMakeFiles/mocap_helper.dir/build.make:209: /home/mobilerobot/Crazyflie_robots/crazyswarm/ros_ws/devel/lib/crazyswarm/mocap_helper] Error 1
make[1]: *** [CMakeFiles/Makefile2:1278: crazyswarm/CMakeFiles/mocap_helper.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
/usr/bin/ld: cannot find -lNatNet
collect2: error: ld returned 1 exit status
make[2]: *** [crazyswarm/CMakeFiles/crazyswarm_server.dir/build.make:209: /home/mobilerobot/Crazyflie_robots/crazyswarm/ros_ws/devel/lib/crazyswarm/crazyswarm_server] Error 1
make[1]: *** [CMakeFiles/Makefile2:1312: crazyswarm/CMakeFiles/crazyswarm_server.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```
I'm not sure what else is missing here to get crazyswarm to compile. Any other modifications that come to mind?

**Comment by whoenig** [comment]
Try adding

```
link_directories(
${LIBMOTIONCAPTURE_LINK_DIR}
)
```

at https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/CMakeLists.txt#L155.

I can look into a proper integration (building by default) towards the end of this week. It's not enabled by default, because this would break the Windows and Mac build of libmotioncapture, but since the crazyswarm only supports Ubuntu, we could enable it there always.

**Comment by The-SS** [comment]
That worked. Thanks!

**Reply by hit-JB** [reply]
Hi!,have you solve the issue that the failing to find NatNet,I have do this by adding the link_directories to the CMakeList.txt but nothing happend,could you please tell me how you did it?

**Reply by The-SS** [reply]
In my case, the NatNetSDKCrossplatform package was listening to the wrong ethernet port (I have two). I simply turn off/unplug the second ethernet port when I want to use Crazyswarm. I do everything else per the documentation.

**Reply by The-SS** [reply]
Also, `optitrack_closed_source` is now an option available directly through `hover_swarm.launch`. Make sure you select that if you want to use the closed-source NatNetSDK from Optitrack (that is part the libmotioncapture package which is a [submodule in crazyswarm](https://github.com/USC-ACTLab/crazyswarm/tree/master/ros_ws/src/crazyswarm/externalDependencies)).

**Reply by hit-JB** [reply]
Thanks for your reply,I'll try it

**Reply by hit-JB** [reply]
Sorry, I don't solve the problem ,I mean that the I failed to compile the crazyswarm package, I don't konw where is wrong i just flow the instruction from the Installation guided and the error is blow just as you 
`[ 98%] Linking CXX executable /home/mobilerobot/Crazyflie_robots/crazyswarm/ros_ws/devel/lib/crazyswarm/crazyswarm_server
[ 98%] Linking CXX executable /home/mobilerobot/Crazyflie_robots/crazyswarm/ros_ws/devel/lib/crazyswarm/mocap_helper
/usr/bin/ld: cannot find -lNatNet
collect2: error: ld returned 1 exit status
make[2]: *** [crazyswarm/CMakeFiles/mocap_helper.dir/build.make:209: /home/mobilerobot/Crazyflie_robots/crazyswarm/ros_ws/devel/lib/crazyswarm/mocap_helper] Error 1
make[1]: *** [CMakeFiles/Makefile2:1278: crazyswarm/CMakeFiles/mocap_helper.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
/usr/bin/ld: cannot find -lNatNet
collect2: error: ld returned 1 exit status
make[2]: *** [crazyswarm/CMakeFiles/crazyswarm_server.dir/build.make:209: /home/mobilerobot/Crazyflie_robots/crazyswarm/ros_ws/devel/lib/crazyswarm/crazyswarm_server] Error 1
make[1]: *** [CMakeFiles/Makefile2:1312: crazyswarm/CMakeFiles/crazyswarm_server.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
`

**Reply by hit-JB** [reply]
Hope you can give me some suggestions

**Reply by whoenig** [reply]
did you update all your submodules (or freshly installed the crazyswarm)? This error should disappear in the latest version (this is also verified automatically during continuous integration tests.)

**Comment by whoenig** [comment]
Great. I also made a proper fix, see https://github.com/USC-ACTLab/crazyswarm/pull/520. This will merge within a couple of days, hopefully.

**Reply by hit-JB** [reply]
Thank you for your recovery. By updating the cmake version to the latest, the problem has been resolved

**Comment by The-SS** [comment]
Regarding this, I wanted to clarify that the package _compiled_ with the closed-source option.
 _However_, I was still having issues with mocap_helper getting the optitrack data. I noticed that mocap_helper was only getting point cloud data when a rigid body was created for the crazyflie; i.e. I would have to create rigid bodies to use Crazyswarm's ICP-based tracking (which makes no sense). Labeled and unlabeled markers were both on in Motive 2.3. When  I would check the cf rigid body, mocap_helper would receive the point cloud data; when I would uncheck it, mocap_helper would stop receiving the point cloud data. 

I switched from `optitrack_closed_source` to the open-source `optitrack` implementation and now it worked properly. I don't know exactly what changed from my last comment in [NatNetSDKCrossplatform issue 3](https://github.com/whoenig/NatNetSDKCrossplatform/issues/3), but the open-source solution seems to be working now.

**Reply by whoenig** [reply]
I could not reproduce this issue on my side (using Motive 3.0, though). I only enable "Unlabeled Markers" and "Rigid Bodies" in the streaming settings and the behavior is as expected for me.

**Reply by whoenig** [reply]
Perhaps you could post the output of `./sampleClient` in https://github.com/whoenig/NatNetSDKCrossplatform so I can see how the datastream is encoded for you?

**Reply by The-SS** [reply]
I can post the output when I run Crazyflies next week. However, I played around with the Motive 2.3 settings while sampleClient and packetClient were running. I observed the following:
- Rigid Bodies:
    - packetClient & sampleClient: 
        - enabled: adds the rigid bodies' poses to a new section
        - disabled: removes the rigid bodies from that section without affecting the other outputs
- Labeled Markers:
    - packetClient:
        - enabled: adds all rigid body markers to the list of "labeled markers"
        - disabled: removes _all_ markers from "labeled markers" (without this on, I don't get any marker data)
    - sampleClient:
        - enabled: adds all rigid body markers to the list of "labeled markers"
        - disabled: removes _all_ markers from "labeled markers" (without this on, I don't get any marker data)
- Unlabeled Markers:
    - packetClient:
        - if "Labeled Markers" is enabled, enabling "Unlabeled Markers" adds all other markers (that don't belong to any assets) to the list of _labeled_ markers
    - sampleClient:
        - if "Labeled Markers" is enabled, enabling "Unlabeled Markers" adds all other markers (that don't belong to any assets) to the list of _unlabeled_ markers
I've had to have labeled markers and unlabeled markers enabled to get data. I've also been keeping bodies enabled.

**Reply by The-SS** [reply]
@whoenig 
I finally got the chance to record the output of `./sampleClient`:
- with "labeled markers" disabled and "unlabeled markers" enabled:
```
Timecode : 00:00:00:78.0
Rigid Bodies [Count=0]
Skeletons [Count=0]
Markers [Count=0]
Force Plate [Count=0]
Device [Count=0]
FrameID : 9464
Timestamp : 78.87
Software latency : 1.36 milliseconds
System latency : 13.70 milliseconds
Total client latency : 14.34 milliseconds (transit time +0.61 ms)
...
```
- with "labeled markers" and "unlabeled markers" enabled:
```
Timecode : 00:00:02:34.0
Rigid Bodies [Count=0]
Skeletons [Count=0]
Markers [Count=18]
Unlabeled Marker [ModelID=0, MarkerID=50021] [size=0.02] [pos=0.03,-1.11,0.02]
Unlabeled Marker [ModelID=0, MarkerID=50022] [size=0.02] [pos=0.00,-1.08,0.02]
Unlabeled Marker [ModelID=0, MarkerID=50023] [size=0.02] [pos=-0.97,-0.58,0.03]
... (I skipped the rest of the markers for brevity) ...
Force Plate [Count=0]
Device [Count=0]
FrameID : 12402
Timestamp : 103.35
Software latency : 0.78 milliseconds
System latency : 12.83 milliseconds
Total client latency : 13.05 milliseconds (transit time +0.20 ms)
...
```
The odd thing is that I have to enable the labeled markers to get _any_ markers at all.
I'm running Motive 2.3 and Ubuntu 20.04.

**Reply by whoenig** [reply]
Very odd indeed. I think the difference between the two applications is that I also consider the "legacy data" (see https://github.com/whoenig/NatNetSDKCrossplatform/blob/master/samples/PacketClient/PacketClient.cpp#L1971-L1983). The official SDK seems to just skip this part. That you need to enable labeled markers seems like a bug of Motive (or the SDK) to me.

**Comment by whoenig** [comment]
Thanks for the feedback! I didn't have access to try it yet with the Crazyswarm, so I only focused on the linking issues. I am re-opening this issue to test (&fix) this on my side.
Glad to hear that you are unblocked for now, though!

**Comment by The-SS** [comment, marked as answer]
So the machine I'm using has two ethernet ports, one for OptiTrack and another for internet.
When the open-source solution was used, crazyswarm was defaulting to the ethernet port hooked up to the internet instead of that connected to OptiTrack. 
When I tested with NatNetSDKCrossplatform, the packetClient (open source) did that as well while the sampleClient (closed source) was defaulting to the ethernet port connected to OptiTrack. Thus, the open source would not get data while the closed source would get data.

For now, I'll just have to turn off the internet connection and use the open-source solution.

**Reply by The-SS** [reply]
@whoenig Just a minor follow-up/summary about this:
- Whenever I have internet access (over wifi or ethernet), crazyswarm is not receiving optitack data. 
- If I simply turn off all networks that have internet access then start up `hover_swarm.launch`, the `/pointCloud` rostopic is populated.
- If I turn on the networks that have internet access _after_  `hover_swarm` starts, Crazyswarm continues to operate normally.

So my solution, for now, is to 1) turn off networks that have internet access, 2) start `hover_swarm`, then 3) turn on any other networks.

I'm not sure if this is due to how my current network connections are set up (optitrack streams back to the switch with the cameras), but this roundabout solution seems to work.

**Reply by whoenig** [reply]
That's a problem just with optitrack, not with optitrack_closed_source, correct?

**Reply by The-SS** [reply]
yes, that happens with optitrack. I haven't used optitrack_closed_source yet.

**Comment by whoenig** [comment]
Interesting finding. I think you would need to change https://github.com/IMRCLab/libmotioncapture/blob/main/src/optitrack.cpp#L317 to use the IP address of the Ethernet adapter that you want to listen to (here: the local IP associated with the Ethernet adapter that connects to your optitrack machine). From what I remember, this is how boost decides which network interface to use. If that's the case, I could add an option to change this value without recompiling.

**Comment by The-SS** [comment]
When using NatNetSDKCrossplatform I tried setting [https://github.com/whoenig/NatNetSDKCrossplatform/blob/b1a68bb13fa3bab92ce584e5e295b80229399922/src/main.cpp#L111](https://github.com/whoenig/NatNetSDKCrossplatform/blob/b1a68bb13fa3bab92ce584e5e295b80229399922/src/main.cpp#L111) to the desired IP but that didn't work. 
I also tried changing the line you noted in libmotioncapture but that didn't work either for me.

**Reply by whoenig** [reply]
In https://github.com/USC-ACTLab/crazyswarm/blob/feature_optitrack_closed_source/ros_ws/src/crazyswarm/launch/mocap_helper.launch#L9 (feature_optitrack_closed_source branch!) there is now a new option. This includes the above setting + one additional line that was suggested by other boost users. Perhaps you can give that a try?

**Reply by The-SS** [reply]
I might be able to try this next week. If I can, I'll let you know how things go.
